README.md
# phpunit.el

[![License GPL 3][badge-license]][LICENSE]

* Master [![MELPA Stable](https://stable.melpa.org/packages/phpunit-badge.svg)](https://stable.melpa.org/#/phpunit) [![Circle CI](https://circleci.com/gh/nlamirault/phpunit.el/tree/master.svg?style=svg)](https://circleci.com/gh/nlamirault/phpunit.el/tree/master) [![Coverage Status](https://coveralls.io/repos/nlamirault/phpunit.el/badge.svg)](https://coveralls.io/r/nlamirault/phpunit.el)
* Develop : [![Melpa Status](https://melpa.org/packages/phpunit-badge.svg)](https://melpa.org/#/phpunit) [![Circle CI](https://circleci.com/gh/nlamirault/phpunit.el/tree/develop.svg?style=svg)](https://circleci.com/gh/nlamirault/phpunit.el/tree/develop) [![Coverage Status](https://coveralls.io/repos/github/nlamirault/phpunit.el/badge.svg?branch=develop)](https://coveralls.io/github/nlamirault/phpunit.el?branch=develop)

Manage the [PHPUnit][] tests from Emacs (>= 24.3)

## Installation

The recommended way to install ``phpunit.el`` is via [MELPA][]:

    M-x package-install phpunit.el

or [Cask][]:

	(depends-on "phpunit.el")


## Usage

### Available commands

These functions are available :
* `phpunit-current-test`: launch unit tests for the current test in a class
* `phpunit-current-class`: launch unit tests for the current class
* `phpunit-current-project`: launch all unit tests
* `phpunit-rerun`: re-launch the most recently launched unit tests 
* `phpunit-group`: launch PHPUnit for group

You can create some key bindings with these commands:

```lisp
(define-key web-mode-map (kbd "C-t t") 'phpunit-current-test)
(define-key web-mode-map (kbd "C-t c") 'phpunit-current-class)
(define-key web-mode-map (kbd "C-t p") 'phpunit-current-project)
(define-key web-mode-map (kbd "C-t r") 'phpunit-rerun)
```

or use the minor mode :

```lisp
(add-to-list 'auto-mode-alist '("\\.php$'" . phpunit-mode))
```



### Configuration

The following configuration variables are available:

```lisp
(setq phpunit-configuration-file "phpunit.xml")
(setq phpunit-root-directory "./")
```

## Development

### Cask

``phpunit.el`` use [Cask](https://github.com/cask/cask) for dependencies
management. Install it and retrieve dependencies :

    $ curl -fsSL https://raw.githubusercontent.com/cask/cask/master/go | python
    $ export PATH="$HOME/.cask/bin:$PATH"
    $ cask

### Testing

* Launch unit tests from shell

        $ make clean test

* Using [overseer][] :

Keybinding           | Description
---------------------|------------------------------------------------------------
<kbd>C-c , t</kbd>   | launch unit tests from buffer
<kbd>C-c , b</kbd>   | launch unit tests
<kbd>C-c , g</kbd>   | launch unit tests with tag (find, regexp, ...)

* Tips:

If you want to launch a single unit test, add a specify tag :

```lisp
(ert-deftest test-foobar ()
  :tags '(current)
  ```

And launch it using : <kbd>C-c , g</kbd> and specify tag : *current*


## Support / Contribute

See [here](CONTRIBUTING.md)



## Changelog

A changelog is available [here](ChangeLog.md).


## License

See [LICENSE](LICENSE).


## Contact

Nicolas Lamirault <nicolas.lamirault@gmail.com>


[badge-license]: https://img.shields.io/badge/license-GPL_2-green.svg?style=flat
[LICENSE]: https://github.com/nlamirault/phpunit.el/blob/master/LICENSE
[Issue tracker]: https://github.com/nlamirault/phpunit.el/issues

[PHPUnit]: http://phpunit.de

[GNU Emacs]: https://www.gnu.org/software/emacs/
[MELPA]: https://melpa.org/
[Cask]: http://cask.github.io/
[Overseer]: https://github.com/tonini/overseer.el


ChangeLog.md
# phpunit.el ChangeLog

## Unreleased

- fix: Improve handling of `hide-compilation-buffer`
- feat: Rerun the most recently run test

## Version 0.17.1 (12/08/2018)

- [#60](https://github.com/nlamirault/phpunit.el/pull/60): Hotfix: Path to `phpunit` could not be expanded

## Version 0.17.0 (09/08/2018)

- [#54](https://github.com/nlamirault/phpunit.el/pull/54): Fix warnings
- [#55](https://github.com/nlamirault/phpunit.el/pull/55): Integrate .dir-locals.el
- [#56](https://github.com/nlamirault/phpunit.el/pull/56): Fix `Makefile` for test
- [#57](https://github.com/nlamirault/phpunit.el/pull/57): Rename variable names
- [#58](https://github.com/nlamirault/phpunit.el/pull/58): Fix for spaces in project path (thanks @jflearn)

## Version 0.16.0 (27/11/2017)

- [#52](https://github.com/nlamirault/phpunit.el/pull/52): windows system prohibits the use of stty (tszg)
- [#49](https://github.com/nlamirault/phpunit.el/pull/49): Add option to hide compilation buffer if all tests pass (thanks mallt)
- [#48](https://github.com/nlamirault/phpunit.el/pull/48): Support colorize output (thanks zonuexe)
- [#46](https://github.com/nlamirault/phpunit.el/pull/46): Add custom variables :tag (thanks zonuexe)
- [#45](https://github.com/nlamirault/phpunit.el/pull/45): Add path to current test file for phpunit-current-test (thanks landakram)
- [#44](https://github.com/nlamirault/phpunit.el/pull/44): Add ability to specify a bootstrap file (thanks landakram)

## Version 0.15.0 (02/11/2017)

- [#43](https://github.com/nlamirault/phpunit.el/pull/42): Rename test-helper to phpunit-test-helper (thanks zonuexe)
- [#41](https://github.com/nlamirault/phpunit.el/pull/41): Remove bundled ert.el (thanks tarsius)

## Version 0.14.0 (12/19/2016)

- [#39](https://github.com/nlamirault/phpunit.el/pull/39): Support custom phpunit command (thanks zonuexe)
- Modify phpunit detection algorithm
- [#36](https://github.com/nlamirault/phpunit.el/pull/36): Modify custom group (thanks zonuexe)

## Version 0.13.0 (03/09/2016)

- [#34](https://github.com/nlamirault/phpunit.el/pull/34): Fix test phpunit-get-class (thanks zonuexe)
- [#33](https://github.com/nlamirault/phpunit.el/pull/33): Fix phpunit-current-test (thanks zonuexe)
- [#30](https://github.com/nlamirault/phpunit.el/pull/30): Modify phpunit command priority (thanks zonuexe)
- [#29](https://github.com/nlamirault/phpunit.el/pull/29): Refactor customize (thanks zonuexe)

## Version 0.12.0 (08/08/2016)

- `FIX` unit tests if *phpunit* executable exists.
- `FIX` Unit tests launcher using *overseer*
- Update documentation for unit testing
- Setup code coverage badges for *master* and *develop* branches

## Version 0.11.0 (08/08/2016)

- [#25](https://github.com/nlamirault/phpunit.el/pull/25): Add phpunit-group (Thanks zonuexe)

## Version 0.10.0 (08/07/2016)

- [#23](https://github.com/nlamirault/phpunit.el/pull/23): Better regexp using rx (Thanks zonuexe)
- [#22](https://github.com/nlamirault/phpunit.el/pull/22): Simplify phpunit-current-class (Thanks zonuexe)
- [#21](https://github.com/nlamirault/phpunit.el/pull/21): Fix for Emacs 25.1 (Thanks zonuexe)

## Version 0.9.0 (05/31/2016)

- [#19](https://github.com/nlamirault/phpunit.el/pull/19): Some problem fixes (Tramp, phpunit-get-root-directory,
  Use phpunit -c option when set configuration file, ...) (thanks zonuexe)
- Remove keybinding from phpunit-helm

## Version 0.8.0 (05/12/2016)

- [#16](https://github.com/nlamirault/phpunit.el/pull/16): Create a minor mode (thanks eric-hansen)
- [#12](https://github.com/nlamirault/phpunit.el/pull/12): Using the phpunit installed by Composer (thanks wangchen)

## Version 0.7.0 (09/10/2015)

- [#10](https://github.com/nlamirault/phpunit.el/pull/10): Set stty columns before running the phpunit command (Thanks Ryckes)

## Version 0.6.0 (08/27/2015)

- [#9](https://github.com/nlamirault/phpunit.el/pull/9): Exact Class Names and Naming Conventions (Thanks muddletoes)

## Version 0.5.0 (08/23/2015)

- [#7](https://github.com/nlamirault/phpunit.el/pull/7) : `php-get-current-class` matches style convention (Thanks muddletoes)
- [#5](https://github.com/nlamirault/phpunit.el/pull/5) : Add error navigation after a failed test (Thanks Ryckes)

## Version 0.4.0 (03/02/2015)

- Update unit tests configuration for [overseer][]
- [#4](https://github.com/nlamirault/phpunit.el/pull/4): Variables for phpunit's directory and filename (Ahmad N. Raja)
- Update [TravisCI][] and [Drone.io][] for continuous integration
- Add code coverage using [undercover][]

## Version 0.3.0 (10/24/2014)

- [#2](https://github.com/nlamirault/phpunit.el/pull/2): Remove unnecessary "depends-on" from Cask
- `FIX` Add Package-Requires line to display dependencies
- Update documentation

## Version 0.2.0 (09/04/2014)

- Launch PHPUnit on current test
- Add optional arguments to PHPUnit
- Use `defcustom` instead of `defvar` for customizable variables
  (Thanks to [Syohex](https://github.com/syohex))


## Version 0.1.0 (04/04/2014)

- Launch current class tests
- Launch all tests from current projects


[TravisCI]: https://travis-ci.org/nlamirault/emacs-travis
[Drone.io]: https://drone.io/github.com/nlamirault/emacs-travis
[overseer]: https://github.com/tonini/overseer.el
[undercover]: https://github.com/sviridov/undercover.el


test/ert-loader.el
;;; ert-loader.el --- Load Ert if not included in Emacs

(require 'f)
(eval-when-compile
  (require 'cl))


(defvar phpunit-root-path
  (f-parent (f-parent load-file-name))
  "Path to root.")

(defvar phpunit-vendor-path
  (f-expand "vendor" phpunit-root-path)
  "Path to vendor.")

(unless (require 'ert nil 'noerror)
  (require 'ert (f-expand "ert" phpunit-vendor-path)))



test/test-helper.el
;; phpunit-test-helper.el --- Test helpers for phpunit.el

;; Copyright (C) 2014-2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; Author: Nicolas Lamirault <nicolas.lamirault@chmouel.com>
;; Homepage: https://github.com/nlamirault/phpunit.el

;;; License:

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ansi)
(require 'cl) ;; http://emacs.stackexchange.com/questions/2864/symbols-function-definition-is-void-cl-macroexpand-all-when-trying-to-instal
(require 'ert)
(require 'f)
(require 'undercover)


(setq debugger-batch-max-lines (+ 50 max-lisp-eval-depth)
      debug-on-error t)

(defvar phpunit-test-helper-username (getenv "HOME"))

(defconst phpunit-testsuite-dir (f-parent (f-this-file))
  "The testsuite directory.")

(defconst phpunit-source-dir
  (f-parent phpunit-testsuite-dir)
  "The Phpunit.el source directory.")

;; (message "Running tests on Emacs %s" emacs-version)

;; (message "Load Phpunit : %s" phpunit-source-dir)
;; (load (s-concat phpunit-source-dir "/phpunit.elc"))

(defconst phpunit-sandbox-path
  (f-expand "sandbox" phpunit-testsuite-dir)
  "The sandbox path for phpunit.")

(defun phpunit-test-helper-cleanup-load-path ()
  "Remove home directory from 'load-path."
  (message (ansi-green "[phpunit] Cleanup path"))
  (mapc #'(lambda (path)
            (when (string-match (s-concat phpunit-test-helper-username "/.emacs.d") path)
              (message (ansi-yellow "Suppression path %s" path))
              (setq load-path (delete path load-path))))
        load-path))

(defun phpunit-test-helper-load-unit-tests (path)
  "Load all unit test from PATH."
  (message (ansi-green "[phpunit] Execute unit tests %s"
                       path))
  (dolist (test-file (or argv (directory-files path t "-test.el$")))
    (load test-file nil t)))

(defun phpunit-test-helper-load-library (file)
  "Load current library from FILE."
  (let ((path (s-concat phpunit-source-dir file)))
    (message (ansi-yellow "[phpunit] Load library from %s" path))
    (undercover "*.el" (:exclude "*-test.el"))
    (require 'phpunit path)))

(defmacro phpunit-test-helper-with-test-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(unwind-protect
       (condition-case nil ;ex
           (let ((default-directory phpunit-source-dir))
             (phpunit-test-helper-cleanup-load-path)
             (phpunit-test-helper-load-library "/phpunit.el")
             ,@body))))

;;; test-helper.el ends here


test/phpunit-test.el
;;; phpunit-test.el --- Tests for phpunit.el

;; Copyright (C) 2014-2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;;; Commentary:

;; Unit tests for phpunit.el

;;; License:

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Code:
(require 'ert)
(require 'f)

;; (defun phpunit-command (&rest arg)
;;   (let ((composer-dir (s-concat (concat (getenv "HOME") "/") ".composer"))
;;         (conf (if phpunit-configuration-file
;;                   (s-concat "-c " phpunit-configuration-file " ")
;;                 "")))
;;     (if (f-dir? composer-dir)
;;         (apply 's-concat composer-dir "/vendor/bin/phpunit " conf arg)
;;       (apply 's-concat "./vendor/bin/phpunit " conf arg))))


(ert-deftest test-phpunit-get-class()
  :tags '(tools)
  (should (string= "PhpUnitTest"
                   (with-temp-buffer
                     (insert "<?php
class PhpUnitTest extends \\PHPUnit_Framework_TestCase {
    public function test() {
    }
}")
                     (goto-char (point-max))
                     (phpunit-get-current-class))))

  (should (string= "PhpUnitTest"
                   (with-temp-buffer
                     (insert "<?php
namespace MyProj;

final class PhpUnitTest extends TestCase
{
    public function test_foo()
    {
    }
}
")
                     (goto-char (point-max))
                     (phpunit-get-current-class))))
  (should (eq nil
              (with-temp-buffer
                (insert "<?php

class PhpUnitTest extends \\PHPUnit_Framework_TestCase {
    public function test() {
    }
}")
                (goto-char (point-min))
                (phpunit-get-current-class)))))

(ert-deftest test-phpunit-get-class()
  :tags '(tools)
  (should (string= "test"
                   (with-temp-buffer
                     (insert "<?php
class PhpUnitTest extends \\PHPUnit_Framework_TestCase {
    public function test() {
    }
}")
                     (goto-char (point-max))
                     (phpunit-get-current-test))))
  (should (string= "test_foo"
                   (with-temp-buffer
                     (insert "<?php
namespace MyProj;

final class PhpUnitTest extends TestCase
{
    public function test_foo()
    {
    }
}
")
                     (goto-char (point-max))
                     (phpunit-get-current-test))))

  (should (eq nil
              (with-temp-buffer
                (insert "<?php
class PhpUnitTest extends \\PHPUnit_Framework_TestCase {
    public function test() {
    }
}")
                (goto-char (point-min))
                (phpunit-get-current-test)))))

;; Using configuration file

(ert-deftest test-phpunit-with-configuration-file ()
  :tags '(configuration-file)
  (phpunit-test-helper-with-test-sandbox
   (let ((phpunit-configuration-file "phpunit.xml"))
     (should (s-contains? (s-concat "-c " (f-long phpunit-configuration-file))
                          (phpunit-get-program (phpunit-arguments "")))))))

(ert-deftest test-phpunit-without-configuration-file ()
  :tags '(configuration-file)
  (phpunit-test-helper-with-test-sandbox
   (should-not (s-contains? "-c phpunit.xml"
                            (phpunit-get-program (phpunit-arguments ""))))))

;; Arguments

(ert-deftest test-phpunit-get-program-without-args ()
  :tags '(arguments)
  (phpunit-test-helper-with-test-sandbox
   (should (s-suffix? "phpunit "
                      (phpunit-get-program (phpunit-arguments ""))))))

(ert-deftest test-phpunit-add-stop-on-error-argument ()
  :tags '(arguments)
  (phpunit-test-helper-with-test-sandbox
   (let ((phpunit-stop-on-error t))
     (should (s-suffix? "phpunit  --stop-on-error"
                        (phpunit-get-program (phpunit-arguments "")))))))

(ert-deftest test-phpunit-add-stop-on-failure-argument ()
  :tags '(arguments)
  (phpunit-test-helper-with-test-sandbox
   (let ((phpunit-stop-on-failure t))
     (should (s-suffix? "phpunit  --stop-on-failure"
                        (phpunit-get-program (phpunit-arguments "")))))))

(ert-deftest test-phpunit-add-stop-on-skipped-argument ()
  :tags '(arguments)
  (phpunit-test-helper-with-test-sandbox
   (let ((phpunit-stop-on-skipped t))
     (should (s-suffix? "phpunit  --stop-on-skipped"
                        (phpunit-get-program (phpunit-arguments "")))))))

(ert-deftest test-phpunit-add-verbose-argument ()
  :tags '(arguments)
  (phpunit-test-helper-with-test-sandbox
   (let ((phpunit-verbose-mode t))
     (should (s-suffix? "phpunit  --verbose"
                        (phpunit-get-program (phpunit-arguments "")))))))


(provide 'phpunit-test)
;;; phpunit-test.el ends here


phpunit-mode.el
;;; phpunit-mode.el --- Minor mode for PHPUnit

;; Copyright (C) 2014-2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;;; Code:

(require 'phpunit)

(defvar phpunit-mode-map
  (let ((map (make-keymap)))
    (define-key map (kbd "C-t t") 'phpunit-current-test)
    (define-key map (kbd "C-t c") 'phpunit-current-class)
    (define-key map (kbd "C-t p") 'phpunit-current-project)
    (define-key map (kbd "C-t r") 'phpunit-rerun)
    map)
  "Keymap for PHPUnit minor mode.")

(define-minor-mode phpunit-mode
  "PHPUnit minor mode"
  :lighter " phpunit"
  :keymap phpunit-mode-map)

(add-to-list 'auto-mode-alist '("\\.php$'" . phpunit-mode))

;;; phpunit-mode.el ends here


phpunit.el
;;; phpunit.el --- Launch PHP unit tests using phpunit

;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;;         Eric Hansen <hansen.c.eric@gmail.com>
;;
;; URL: https://github.com/nlamirault/phpunit.el
;; Version: 0.17.1
;; Keywords: tools, php, tests, phpunit

;; Package-Requires: ((s "1.12.0") (f "0.19.0") (pkg-info "0.6") (cl-lib "0.5") (emacs "24.3"))

;;; License:

;; Copyright (C) 2014-2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;; Thanks to tox.el(https://github.com/chmouel/tox.el) from Chmouel Boudjnah.

;; To use this code, bind the functions `phpunit-current-test', `phpunit-current-class',
;; and `phpunit-current-project' to convenient keys with something like :

;; (define-key web-mode-map (kbd "C-x t") 'phpunit-current-test)
;; (define-key web-mode-map (kbd "C-x c") 'phpunit-current-class)
;; (define-key web-mode-map (kbd "C-x p") 'phpunit-current-project)

;;; Code:

(require 'cl-lib)
(require 'compile)
(require 'tramp)
(require 's)
(require 'f)
(eval-when-compile
  (require 'rx))

(defgroup phpunit nil
  "PHPUnit utility"
  :tag "PHPUnit"
  :prefix "phpunit-"
  :group 'tools
  :group 'php)

(defcustom phpunit-default-program nil
  "PHPUnit command or path to executable file or a function that returns these string."
  :type '(choice (file     :tag "Path to PHPUnit executable file.")
                 (function :tag "A function return PHPUnit executable file path.")
                 (string   :tag "PHPUnit command name. (require command in PATH)")))

(defvar phpunit-program)
(make-obsolete-variable 'phpunit-program 'phpunit-default-program "0.18.0")

(defvar phpunit-arg)
(make-obsolete-variable 'phpunit-arg 'phpunit-args "0.18.0")

(defcustom phpunit-stop-on-error nil
  "Stop execution upon first error."
  :type 'boolean)

(defcustom phpunit-stop-on-failure nil
  "Stop execution upon first error or failure."
  :type 'boolean)

(defcustom phpunit-stop-on-skipped nil
  "Stop execution upon first skipped test."
  :type 'boolean)

(defcustom phpunit-verbose-mode nil
  "Display debugging information during test execution."
  :type 'boolean)

(defcustom phpunit-configuration-file nil
  "The PHPUnit configuration file."
  :type '(choice (file  :tag "Path to phpunit.xml[.dist]")
                 (const :tag "Automatically detect the path of phpunit.xml" nil)))

(defcustom phpunit-bootstrap-file nil
  "The PHPUnit bootstrap file."
  :type '(choice (file  :tag "Path to PHPUnit bootstrap script")
                 (const :tag "Not specify boostrap script" nil)))

(defcustom phpunit-colorize nil
  "Colorize PHPUnit compilation output buffer."
  :type '(choice (const :tag "Do not specific --color argument" nil)
                 (const :tag "--color=auto" "auto")
                 (const :tag "--color=never" "never")
                 (const :tag "--color=always" "always")))

(defcustom phpunit-hide-compilation-buffer-if-all-tests-pass nil
  "Hide the compilation buffer if all tests pass."
  :type 'boolean
  :group 'phpunit)

(defconst php-beginning-of-defun-regexp
  (eval-when-compile
    (rx line-start
        (* (syntax whitespace))
        (* (or "abstract" "final" "private" "protected" "public" "static") (+ (syntax whitespace)))
        (* (syntax whitespace))
        "function"
        (+ (syntax whitespace))
        (? "&")
        (group (+ (or (syntax word) (syntax symbol))))
        (* (syntax whitespace))
        "("))
  "Regular expression for a PHP function.")

(defconst php-beginning-of-class
  (rx line-start
        (* (syntax whitespace))
        (? "final" (syntax whitespace))
        (* (syntax whitespace))
        "class"
        (+ (syntax whitespace))
        (group (+ (or (syntax word) (syntax symbol))))
        (* (syntax whitespace)))
  "Regular expression for a PHP class.")

(defconst php-labelchar-regexp
  "[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]"
  "Valid syntax for a character in a PHP label.")

;;;###autoload
(progn
  (defvar-local phpunit-root-directory nil
    "Directory path to execute PHPUnit.")
  (put 'phpunit-root-directory 'safe-local-variable #'stringp)
  (defvar-local phpunit-args nil
    "Argument to pass to phpunit command.")
  (put 'phpunit-args 'safe-local-variable #'(lambda (v) (or (stringp v) (listp v))))
  (defvar-local phpunit-executable nil
    "PHPUnit command or path to executable file.")
  (put 'phpunit-executable 'safe-local-variable
       #'(lambda (v) (or (null v) (stringp v) (functionp v)))))

(when phpunit-hide-compilation-buffer-if-all-tests-pass
  (add-hook 'compilation-finish-functions 'phpunit--hide-compilation-buffer-if-all-tests-pass))

(defvar phpunit-last-group-cache nil)

;; Commands
;; -----------

(defun phpunit--find-executable (directory)
  "Get PHPUnit executable command in `DIRECTORY'."
  (let ((executable (or phpunit-executable phpunit-default-program)))
    (cond ((stringp executable) executable)
          ((functionp executable) (funcall executable))
          ((and directory
                (file-exists-p (concat directory "vendor/bin/phpunit")))
           (expand-file-name (concat directory "vendor/bin/phpunit")))
          ((executable-find "phpunit") "phpunit")
          (t (error "PHPUnit command/package is not installed")))))

(defun phpunit-get-program (args)
  "Return the command to launch unit test.
`ARGS' corresponds to phpunit command line arguments."
  (let* ((vendor-dir (locate-dominating-file default-directory "vendor"))
         (executable (phpunit--find-executable vendor-dir)))
    (when (file-remote-p default-directory)
      (setq executable
            (tramp-file-name-localname (tramp-dissect-file-name executable))))
    (s-concat (shell-quote-argument executable)
              (when phpunit-args
                (s-concat " " (if (stringp phpunit-args) phpunit-args
                                (s-join " " (mapcar 'shell-quote-argument phpunit-args)))))
              (if phpunit-configuration-file
                  (s-concat " -c " (shell-quote-argument (expand-file-name phpunit-configuration-file)))
                "")
              (if phpunit-bootstrap-file
                  (s-concat " --bootstrap " (shell-quote-argument (expand-file-name phpunit-bootstrap-file)))
                "")
              (when phpunit-colorize
                (format " --colors=%s" phpunit-colorize))
              " "
              args)))

(defun phpunit-get-root-directory (&optional force-real-path)
  "Return the root directory to run tests."
  ;; The function doesn't detect the root directory when used with
  ;; tramp mode. In that case, the phpunit-root-directory variable can
  ;; be set which takes precedence
  (or (and (not force-real-path) phpunit-root-directory)
      (let ((filename (buffer-file-name)) path)
        (cond
         ((null filename) default-directory)
         (phpunit-configuration-file
          (file-truename (locate-dominating-file filename phpunit-configuration-file)))
         (:else
          (cl-loop for file in '("phpunit.xml" "phpunit.xml.dist" ".git" "composer.json")
                   do (setq path (locate-dominating-file filename file))
                   if path return (file-truename path)
                   finally return (file-truename "./")))))))

(defun phpunit-get-current-class ()
  "Return the canonical unit test class name associated with the current class or buffer."
  (save-excursion
    (when (re-search-backward php-beginning-of-class nil t)
      (match-string-no-properties 1))))

(defun phpunit-get-current-test ()
  "Get the name of the current test function"
  (save-excursion
    (when (re-search-backward php-beginning-of-defun-regexp nil t)
      (match-string-no-properties 1))))

(defun phpunit--listing-groups ()
  "Return list of @group.

https://phpunit.de/manual/current/en/appendixes.annotations.html#appendixes.annotations.group"
  (let ((phpunit-output (phpunit--execute "--list-groups")))
    (with-temp-buffer
      (insert phpunit-output)
      (goto-char (point-min))
      (search-forward "Available test group")
      (move-beginning-of-line 1)
      (forward-line)
      (cl-loop
       for line in (s-split "\n" (buffer-substring-no-properties (point) (point-max)))
       if (s-starts-with? " - " line)
       collect (s-chop-prefix " - " line)))))

(defun phpunit--get-last-group (path)
  "Get last group cache by `PATH'."
  (if (null phpunit-last-group-cache)
      nil
    (gethash path phpunit-last-group-cache nil)))

(defun phpunit--put-last-group (group path)
  "Put last group `GROUP' cache by `PATH'."
  (unless phpunit-last-group-cache
    (setq phpunit-last-group-cache (make-hash-table :test 'equal)))
  (puthash path group phpunit-last-group-cache))

(defun phpunit-arguments (args)
  "Append options to `ARGS' by variables."
  (when phpunit-stop-on-error
    (setq args (s-concat args " --stop-on-error")))
  (when phpunit-stop-on-failure
    (setq args (s-concat args " --stop-on-failure")))
  (when phpunit-stop-on-skipped
    (setq args (s-concat args " --stop-on-skipped")))
  (when phpunit-verbose-mode
    (setq args (s-concat args " --verbose")))
  args)

(defun phpunit-get-compile-command (args)
  "Return command string to execute PHPUnit from `ARGS'."
  (if (memq system-type '(windows-nt ms-dos))
      (phpunit-get-program (phpunit-arguments args))
    (let ((column-setting-command (format "stty cols %d" (frame-width)))
	  (command-separator "; ")
	  (phpunit-command (phpunit-get-program (phpunit-arguments args))))
      (concat column-setting-command command-separator phpunit-command))))

(defun phpunit--colorize-compilation-buffer ()
  "Colorize PHPUnit compilation buffer."
  (let ((inhibit-read-only t))
    (ansi-color-apply-on-region compilation-filter-start (point))))

(defun phpunit--setup-compilation-buffer ()
  "Setup hooks for PHPUnit compilation buffer."
  (add-hook 'compilation-finish-functions #'phpunit--finish-compilation-buffer)
  (add-hook 'compilation-filter-hook #'phpunit--colorize-compilation-buffer))

(defun phpunit--finish-compilation-buffer (&optional cur-buffer msg)
  "Setup hooks for PHPUnit compilation buffer.
`CUR-BUFFER' and `MSG' are ignore."
  (remove-hook 'compilation-finish-functions #'phpunit--finish-compilation-buffer)
  (remove-hook 'compilation-filter-hook #'phpunit--colorize-compilation-buffer))

(defun phpunit--execute (args)
  "Execute phpunit command with `ARGS'."
  (let ((default-directory (phpunit-get-root-directory)))
    (shell-command-to-string (phpunit-get-program (phpunit-arguments args)))))

(defun phpunit-run (args &optional dir)
  "Execute phpunit command with `ARGS' in `DIR'.

If `DIR' is not provided, it will be inferred with phpunit-get-root-directory."
  (add-to-list 'compilation-error-regexp-alist '("^\\(.+\\.php\\):\\([0-9]+\\)$" 1 2))
  (let ((default-directory (or dir (phpunit-get-root-directory)))
        (compilation-process-setup-function #'phpunit--setup-compilation-buffer))
    (compile (phpunit-get-compile-command args))))

;; This function will be overwritten each time a test is run.
(defun phpunit--run-last ()
  "Rerun the last phpunit command."
  (error "Nothing to rerun: try again after running a test"))

(defun phpunit--hide-compilation-buffer-if-all-tests-pass (buffer status)
  "Hide the compilation BUFFER if all tests pass.
The STATUS describes how the compilation process finished."
  (with-current-buffer buffer
    (let* ((buffer-string (buffer-substring-no-properties
                           (point-min) (point-max)))
           (buffer-lines (s-lines buffer-string))
           (ok-msg (car (cl-remove-if-not
                         (lambda (x)
                           (and (s-contains? "OK" x)
                                (s-contains? "test" x)
                                (s-contains? "assertion" x)))
                         buffer-lines)))
           (time-msg (car (cl-remove-if-not
                           (lambda (x)
                             (and (s-contains? "Time" x)
                                  (s-contains? "Memory" x)))
                           buffer-lines))))
      (when ok-msg
        (delete-windows-on buffer)
        (message "%s %s" ok-msg time-msg)))))

;; API
;; ----

;;;###autoload
(defun phpunit-set-dir-local-variable (variable)
  "Create project file `.dir-locals.el' and set `VARIABLE' for `phpunit.el'."
  (interactive
   (list (intern (completing-read "Select variable name: "
                                  '(phpunit-executable phpunit-directory)))))
  (add-dir-local-variable nil variable
                          (read-string
                           (cl-case variable
                             ('phpunit-executable "Input a command to run PHPUnit: ")
                             ('phpunit-root-directory "Input a directory path to execute PHPUnit: ")))))

;;;###autoload
(defun phpunit-current-test ()
  "Launch PHPUnit on curent test."
  (interactive)
  (let* ((args (s-concat " --filter '"
                         (phpunit-get-current-class)
                         "::"
                         (phpunit-get-current-test) "'"
                         " "
                         (s-chop-prefix (phpunit-get-root-directory) buffer-file-name)))
         (cmd (apply-partially #'phpunit-run args (phpunit-get-root-directory))))
    (fset #'phpunit--run-last cmd)
    (funcall cmd)))

;;;###autoload
(defun phpunit-current-class ()
  "Launch PHPUnit on current class."
  (interactive)
  (let ((cmd (apply-partially #'phpunit-run (s-chop-prefix (phpunit-get-root-directory t) buffer-file-name) (phpunit-get-root-directory))))
    (fset #'phpunit--run-last cmd)
    (funcall cmd)))

;;;###autoload
(defun phpunit-current-project ()
  "Launch PHPUnit on current project."
  (interactive)
  (let ((cmd (apply-partially #'phpunit-run "" (phpunit-get-root-directory))))
    (fset #'phpunit--run-last cmd)
    (funcall cmd)))

;;;###autoload
(defun phpunit-group (use-last-group &optional group)
  "Launch PHPUnit for group."
  (interactive "p")
  (let* ((current-root-directory (phpunit-get-root-directory))
         (last-group (phpunit--get-last-group current-root-directory)))
    (when (called-interactively-p 'interactive)
      (setq use-last-group (eq use-last-group 1))
      (setq group (if (and use-last-group last-group)
                      last-group
                    (completing-read "PHPUnit @group: " (phpunit--listing-groups)))))
    (phpunit-run (format "--group %s" group))
    (phpunit--put-last-group group current-root-directory)))

;;;###autoload
(defun phpunit-rerun ()
  "Rerun the last PHPUnit test."
  (interactive)
  (phpunit--run-last))

(provide 'phpunit)
;;; phpunit.el ends here
